### PR TITLE
feat: add a new shim for lunatic

### PR DESCRIPTION
**Pull request by 0xE282B0**
Shim for [Lunatic](https://lunatic.solutions/) "an Erlang-inspired runtime for WebAssembly".

### CODE REVIEW

**Comment by 0xE282B0** [comment]
@microsoft-github-policy-service agree company="Liquid Reply"

File: containerd-shim-lunatic-v1/src/executor.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        Some(p) => p,
+    };
+
+    match p.args() {
+        None => &[],
+        Some(args) => args.as_slice(),
+    }
+}
+
+impl Executor for LunaticExecutor {
```
Comment (Mossaka) [inline containerd-shim-lunatic-v1/src/executor.rs:52]: I like that it's already using libcontainer!!

**Comment by Mossaka** [comment]
Ready for review?

**Comment by 0xE282B0** [comment]
@Mossaka, I wanted to use `LibcontainerInstance` but can't make it build.

When I enable the `libcontainer` feature on `containerd-shim-wasm` it adds `libcontainer/default` features. With that I need  to add th packages `libdbus-1-dev` and `libsystemd0` finally I get the error:

```
/usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: 
    warning: libpthread.so.0, needed by /usr/lib/x86_64-linux-gnu/libdbus-1.so, not found (try using -rpath or -rpath-link)
```
and
```
/usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: /usr/lib/x86_64-linux-gnu/libdbus-1.so: undefined reference to `sd_listen_fds@LIBSYSTEMD_209'
/usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: /usr/lib/x86_64-linux-gnu/libdbus-1.so: undefined reference to `sd_is_socket@LIBSYSTEMD_209'
/usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: /usr/lib/x86_64-linux-gnu/libdbus-1.so: undefined reference to `pthread_mutexattr_settype@GLIBC_2.2.5'
/usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: /usr/lib/x86_64-linux-gnu/libdbus-1.so: undefined reference to `pthread_mutexattr_init@GLIBC_2.2.5'
/usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: /usr/lib/x86_64-linux-gnu/libdbus-1.so: undefined reference to `pthread_mutexattr_destroy@GLIBC_2.2.5'
/usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: /usr/lib/x86_64-linux-gnu/libdbus-1.so: undefined reference to `pthread_condattr_setclock@GLIBC_2.3.3'
```

Does the `libcontainer` feater really need to depend on systemd and dbus or am I doing something completely wrong?

**Comment by 0xE282B0** [comment]
When I remove the default features on `libcontainer` cross compilation works:
https://github.com/containerd/runwasi/commit/00dca5c5c1725da40a528e3a2f938100fac95687

@Mossaka, do we need systemd in libcontainer? Is there a way to cross compile with systemd/dbus dependencies?

**Comment by Mossaka** [comment]
@0xE282B0 can you try "cgroupsv2" feature? 

> @Mossaka, do we need systemd in libcontainer? 

No, musl doesn't have systemd.

**Comment by Mossaka** [comment]
@0xE282B0 I have a PR https://github.com/containerd/runwasi/pull/246 in runwasi to fix this issue. Once it's merged, you can use `cgroupsv2` feature to remove dependencies on `systemd` and `libseccump`.  This should make it compilable against musl.

**Comment by jprendes** [comment]
FYI, you can build with the `systemd` feature in alpine, but:
*  you have to vendor `dbus`, see [here](https://github.com/containerd/runwasi/blob/main/crates/containerd-shim-wasmedge/Cargo.toml#L35) and [here](https://github.com/containerd/runwasi/blob/main/crates/containerd-shim-wasmedge/Cargo.toml#L23)
* if you want to statically link `libseccomp` you need a few env-vars, see [here](https://github.com/containerd/runwasi/blob/main/scripts/dockerfile-utils.sh#L32-L33)

**Comment by jprendes** [comment]
The `systemd` feature in youki doesn't link agains `libsystemd`. It links agains `dbus` instead.
The `dbus` crate can be vendored without involving `systemd`. I do not know if that actually does something in systems without `systemd`

**Comment by Mossaka** [comment]
@0xE282B0 I pushed a new commit that updates to the latest main branch of runwasi. It should build now.

File: Makefile [hunk 2]
```diff
@@ -1,8 +1,9 @@
-SHIMS := slight spin wws
+SHIMS := slight spin wws lunatic
```
Comment (Mossaka) [inline Makefile:1]: Could you please add lunatic to `.github/workflows/release.yaml`?

**Review by Mossaka — COMMENTED** [review]
LGTM! 

I am going to test this locally and then if everything runs fine, will approve this PR.

File: Makefile (hunk 2 above)
Comment (Mossaka) [inline Makefile:1]: Also add `lunatic` to `deployments/k3d/Makefile`

**Comment by Mossaka** [comment]
I am running into this problem from the containerd log:

```
[ERROR] executable "/" does not have the correct permission set
[ERROR] failed to initialize container process: invalid executable: /
```

I think the issue was the libcontainer requires the entrypoint to be an executable, which we tried to push off. See https://github.com/containers/youki/issues/2220.

In the meantime, you probably want to do something like the spin shim is doing described in the https://github.com/deislabs/containerd-wasm-shims/issues/120.

**Comment by Mossaka** [comment]
I just pushed a new commit to make lunatic wasm binaries executable. 

`make integration-tests` is working on my local!!

**Comment by Mossaka** [comment]
Actually, all the pods are running fine, but the response I got is a bit off. Is this expected?

```
curl http://localhost:8082/lunatic
<h1>404: Not found</h1>
```

**Comment by 0xE282B0** [comment]
The path is like spin and wws and the text is just the submillisecond example. What would you expect?

**Comment by Mossaka** [comment]
> The path is like spin and wws and the text is just the submillisecond example. What would you expect?

Thanks for the clarification! I thought we are running the `images/lunatic` example which confuses me. Now that makes sense. 

> Ah, the integration test URL did not match.

Yes, I just pushed a commit to fix it.

**Comment by Mossaka** [comment]
Could you please sign this commit ea3fd9c11bfdfb9e02fa134ba95e1994f8065b0d?

**Comment by Mossaka** [comment]
@0xE282B0 in PR #125, I am trying to fix the CI with splitting build and integration tests into two jobs and thus reduce the disk pressure issue this PR is facing.

**Comment by Mossaka** [comment]
Merging it in!! Thanks @0xE282B0 again for contributing. 🥂

**Pull request merged**